mlfinlab/feature_importance/importance.py

```diff
@@ -0,0 +1,144 @@
+"""
+Module which implements feature importance algorithms described in Chapter 8
+"""
+
+import pandas as pd
+import numpy as np
+from sklearn.metrics import log_loss, accuracy_score, precision_score, f1_score, recall_score, roc_auc_score
+import matplotlib.pyplot as plt
+from mlfinlab.cross_validation.cross_validation import ml_cross_val_score
+
+
+# pylint: disable=invalid-name
+# pylint: disable=E1130
+
+def feature_importance_mean_imp_reduction(clf, feature_names):
+    """
+    Snippet 8.2, page 115. MDI Feature importance
+
+    This function generates feature importance from classifiers estimators importance
+    using Mean Impurity Reduction (MDI) algorithm
+
+    :param clf: (BaggingClassifier, RandomForest or any ensemble sklearn object): trained classifier
+    :param feature_names: (list): array of feature names
+    :return: (pd.DataFrame): individual MDI feature importance
+    """
+    # Feature importance based on IS mean impurity reduction
+    feature_imp_df = {i: tree.feature_importances_ for i, tree in enumerate(clf.estimators_)}
+    feature_imp_df = pd.DataFrame.from_dict(feature_imp_df, orient='index')
+    feature_imp_df.columns = feature_names
+    feature_imp_df = feature_imp_df.replace(0, np.nan)  # Because max_features = 1
+    imp = pd.concat({'mean': feature_imp_df.mean(), 'std': feature_imp_df.std() * feature_imp_df.shape[0] ** -.5},
+                    axis=1)
+    imp /= imp['mean'].sum()
+    return imp
+
+
+def feature_importance_mean_decrease_accuracy(clf, X, y, cv_gen, sample_weight=None, scoring='neg_log_loss'):
+    """
+    Snippet 8.3, page 116-117. MDA Feature Importance
+
+    :param clf: (sklearn.ClassifierMixin): any sklearn classifier
+    :param X: (pd.DataFrame): train set features
+    :param y: (pd.DataFrame, np.array): train set labels
+    :param cv_gen: (PurgedKFold): cross-validation object
+    :param sample_weight: (np.array): sample weights, if None equal to ones
+    :param scoring: (str): scoring function used to determine importance, either 'neg_log_loss' or 'accuracy'
+    :return: (pd.DataFrame): mean and std feature importance
+    """
+    # Feature importance based on OOS score reduction
+    scoring_func_dict = {'neg_log_loss': log_loss, 'accuracy': accuracy_score, 'f1': f1_score,
+                         'precision': precision_score, 'recall': recall_score, 'roc_auc': roc_auc_score}
+    try:
+        scoring_func_dict[scoring]
+    except KeyError:
+        raise ValueError('wrong scoring method.')
```

Comment:
Regarding the error message, could you capitalize the first letter of the first word, and change the wording to point the user to a possible solution? Something like: "Invalid choice of scoring method. Argument 'scoring_method' must be one of: {scoring_func_dict.keys()}'.

Reply:
Done